### ISSUE TITLE
Steuern werden nicht der richtigen Datenreihe zugeordnet

### ISSUE THREAD

**Issue by sekrause** [buchen/portfolio#859]
Beispiel anhand des [angehängten Beispielportfolios](https://github.com/buchen/portfolio/files/1426540/bug.xml.gz) mit einem ETF: Ein Kauf, eine Dividendenausschüttung mit Steuern.

In der Performance-Berechnung des Gesamtportfolios tauchen die Steuern auch auf:

![screen shot 2017-10-30 at 09 56 21](https://user-images.githubusercontent.com/595780/32162476-64289cfe-bd59-11e7-89e9-649059075b97.png)

Wenn ich als Datenreihe nun den risikobehafteten Portfolioanteil auswähle, werden die Steuern hier nicht mehr angezeigt:

![screen shot 2017-10-30 at 09 56 51](https://user-images.githubusercontent.com/595780/32162526-9fa3d4a6-bd59-11e7-9775-0eace4e9c95f.png)

Da die Steuern ja bei einer Auszahlung des risikobehafteten Portfolioanteils abgeführt wurden, sollten die hier auch bei der Berechnung angezeigt werden.

**Comment by buchen** [comment]
Ja. Allerdings ist die Logik momentan so implementiert: bei einer Klassifikation werden keine Steuern betrachtet. Die Idee war: wenn ich mehrere Klassifikationen vergleiche, dann stören die Steuern z.B. wenn ich Anfang des Jahres noch einen Freibetrag habe, aber am Ende nicht mehr.

Es wäre schön wenn man das frei wählen könnte - also selber entscheiden könnte, ob eine Klassifikation mit oder ohne Steuern gerechnet wird. Ist auf der List, aber nix kurzfristiges.

**Comment by buchen** [comment]
@imperatom - Welchen OS Version brauchst Du (mac, win64)? (Ich habe Dein Posting im Forum gesehen). Ich habe schon eine Version mit der ich Steuern nicht berücksichtige. Allerdings scheint mir die  Berechnung der Volatilität nicht zu stimmen.

**Comment by imperatom** [comment]
> 
> 
> @imperatom - Welchen OS Version brauchst Du (mac, win64)? (Ich habe Dein Posting im Forum gesehen). Ich habe schon eine Version mit der ich Steuern nicht berücksichtige. Allerdings scheint mir die Berechnung der Volatilität nicht zu stimmen - da brauche ich noch mehr Zeit.

Danke für die Rückmeldung, Andreas. Ich verwende Windows 7. Komisch, dass die Vola-Berechnung nicht hinhaut, die Formel bleibt ja grundsätzlich die gleiche, nur muss sie ausschließlich auf "steuerfreie" Variablen zugreifen. Bei Aktien scheint das ja hinzuhauen - zumindest wird es im Vola-Diagramm aktuell für Aktien steuerfrei angezeigt. 

Durch die aktuell noch unterschiedlichen Methodiken entsteht übrigens folgendes, völlig skuriles Vola-Diagramm, wenn man nur einmal eine einzige Aktie kauft, deren Wert nur einmal von 100 auf 150 (nach Gebühren: 145) gestiegen ist:
![absurdes beispiel](https://user-images.githubusercontent.com/31973162/47443100-ef608c80-d7b3-11e8-8870-a0a445570069.jpg)

Es entstehen verschiedene Punkte für Depot und Aktie, wobei die Methodik jeweils nicht ersichtlich ist und es die einzige Aktie im Depot war.

**Comment by buchen** [comment]
@imperatom https://www.dropbox.com/s/tcnzm8m5ss5jo7l/PortfolioPerformance-distro-0.35.1-SNAPSHOT-win32.win32.x86_64.zip?dl=0 - einfach irgendwo auspacken. Als Datenreihen kann man dann " (ohne Steuern)" auch hinzufügen.

Was mich überrascht hat, dass die Volatilität zwischen vor/nach Steuern so unterschiedlich war. Meinem Bauchgefühl her hätte das nicht sein dürfen. Ich habe es mir aber noch nicht näher angeschaut. Ich weiß bei den Konten habe ich noch ein Problem - da sollte sie ja eigentlich immer 0 sein (es sei denn man hat ein Fremdwährungskonto), es werden aber Zinsen als Bewertungsänderung herangezogen. Und weil ich da mehr Ruhe und Zeit brauche, bin ich einfach noch nicht dazugekommen.

**Comment by imperatom** [comment]
> @imperatom https://www.dropbox.com/s/tcnzm8m5ss5jo7l/PortfolioPerformance-distro-0.35.1-SNAPSHOT-win32.win32.x86_64.zip?dl=0 - einfach irgendwo auspacken. Als Datenreihen kann man dann " (ohne Steuern)" auch hinzufügen.

Nach dem Entpacken sehe ich die veränderte Versionsnummer 0.35.1.201810241823, ansonsten fallen mir aber inhaltlich tatsächlich keine Unterschiede auf. Eine Auswahl "ohne Steuern" kann ich nirgens finden, wenn ich eine Datenreihe zum Diagramm hinzufüge.

**Comment by buchen** [comment]
>  nach dem Entpacken sehe ich die veränderte Versionsnummer 0.35.1.201810241823, ansonsten fallen mir aber inhaltlich tatsächlich keine Unterschiede auf.

Dann habe ich wohl versehentlich den falschen Branch gebaut. Das muss ich mir zu Hause anschauen. Ich kann Samstag noch einen neuen Build hochladen.

**Comment by buchen** [comment]
@imperatom Wenn Du Datenreihen z.B. dem Performance-Diagramm hinzufügst, dann sollte es auch eine Datenreihe "vor Steuern" geben. Das funktioniert auch auf dem Dashboard. Und mit Vola/Rendite Diagramm. Das sollte auch in dem Build drin sein, den ich auf Dropbox geladen habe.

<img width="426" alt="bildschirmfoto 2018-10-27 um 06 53 12" src="https://user-images.githubusercontent.com/587976/47599757-06ba9800-d9b5-11e8-85f8-537da3f72f85.png">

**Comment by buchen** [comment]
> Ich hatte erst nach einer Checkbox gesucht und dann nach Datenreihen für das "Gesamtportfolio" und für andere unter "Wertpapiere" angelegte Portfolioteilgruppen.

Ja, eine Checkbox gefällt mir auch besser. Allerdings muss ich da an den "Innereien" noch umbauen damit das einfacher wird. Dann sind es auch weniger Datenreihen.

> Bei Gelegenheit wäre natürlich toll, das ganze auch mit dem Gesamtportfolio oder Portfolioteilgruppen machen zu können. 

Mach ich. Der Code ist zumindest so dass ich die Berechnung relativ einfach auf andere Berechnungen anwenden kann. Jetzt muss ich noch die Klassifikationen und Wertpapiere geradestehen. Bei ersteren wird ja immer ohne Steuern gerechnet...

**Comment by imperatom** [comment]
@buchen:
Im Umgang mit der Testversion habe ich einen größeren Bug gefunden. Das folgende Bild zeigt wohl schon dessen Ursache, nämlich dass bei Verkäufen die Steuern mit einem DEPOT verknüpft werden, bei Dividenden und separat erfassten Steuern und Steuerrückerstattungen aber nicht:
 
![steuertest](https://user-images.githubusercontent.com/31973162/48777993-5c444500-ecd4-11e8-9725-31179211d889.jpg)

Entsprechend bleiben dann tatsächlich auch bei der Renditeberechnung im Performance-Diagramm nur Verkaufssteuern unberücksichtigt; Die anderen Steuern werden aber fälschlicherweise für die Performance mitberücksichtigt. Daher ist die Performance "ohne Steuern" in Wahrheit nur eine Performance "ohne Verkaufssteuern, aber mit allen anderen Steuern".

**Comment by buchen** [comment]
@imperatom 

> Im Umgang mit der Testversion habe ich einen größeren Bug gefunden. Das folgende Bild zeigt wohl schon dessen Ursache, nämlich dass bei Verkäufen die Steuern mit einem DEPOT verknüpft werden, bei Dividenden und separat erfassten Steuern und Steuerrückerstattungen aber nicht

In der **Performance-Berechnung** kann man aktuell keine Datenreihe *ohne* Steuern auswählen. Ich habe jetzt mal eine Konfiguration hinzugefügt. Wenn man den Filter "vor Steuern" setzt, finden sich keine Buchungen mehr in der Liste *Steuern*. Das ist aber ein guter Hinweis - so kann man die Berechnung viel besser nachvollziehen.

Ich plane im Laufe des Tages (endlich) mal wieder einen neue Version zu veröffentlichen.

**Comment by imperatom** [comment]
> _In der **Performance-Berechnung** kann man aktuell keine Datenreihe _ohne_ Steuern auswählen. Ich habe jetzt mal eine Konfiguration hinzugefügt. Wenn man den Filter "vor Steuern" setzt, finden sich keine Buchungen mehr in der Liste _Steuern_. Das ist aber ein guter Hinweis - so kann man die Berechnung viel besser nachvollziehen._




> Ich plane im Laufe des Tages (endlich) mal wieder einen neue Version zu veröffentlichen.

**Comment by buchen** [comment]
> Ich meinte auch das Performance-Diagramm. In der Testversion, die Du hier oben am 24.10. gepostet hast, konnte man da doch Depots ohne Steuern auswählen. Dabei tritt aber oben beschriebener Fehler auf ("bei der Renditeberechnung im Performance-Diagramm [werden] nur Verkaufssteuern unberücksichtigt", andere Steuern fließen weiterhin ein).

Dann hat mich wohl der Screenshot auf die falsche Fährte gebracht...

Die Berechnung "vor Steuern" ist ein Filter, d.h. ich baue im Speicher eine komplett neue Datei zusammen und schiebe die durch die Berechnungen. Darum gehe ich davon aus, wenn ich die Steuern bei der *Berechnung* nicht sehe, dass sie auch im *Diagramm* nicht vorhanden sind. Selber Code.

Kannst Du das Problem an der *kommer* Beispieldatei nachstellen? Ich habe mir im Performance-Diagramm die letzten drei Monate anzeigen lassen, zwei Datenreihen: einmal das Depot und das Depot vor Steuern. Die Performance unterscheidet sich. Wenn ich die Dividende aus dem November lösche (hat ein paar Steuern), dann ist die Performance identisch.

**Comment by imperatom** [comment]
Hallo Andreas, während Weihnachten ist unser Dialog hier bei mir leider in Vergessenheit geraten - aber jetzt ist er mir wieder eingefallen als ich nämlich erneut mit dem beschriebenen Steuer-Problem konfrontiert war. Die kommer.xml habe ich nicht und finde sie auch nicht im Download-Bereich oder sonst irgendwo hier - wo bekomme ich sie? Stattdessen hier eine Test-Datei, anhand der das Problem leicht deutlich wird:
[test.zip](https://github.com/buchen/portfolio/files/2777402/test.zip)

Es wurde nur mit einem Wertpapier gehandelt, vor Steuern 100% Kursgewinn, nach Steuern 50% Kursgewinn. Zwischenzeitlich gab es eine Dividende, auf die ebenfalls Steuern anfielen.

Der Gesamtportfolio-Chart unter Performance->Diagramm und die Berechnung nach Steuern sind korrekt. Der dort angezeigte Vorsteuer-Graph ist aber falsch. Richtig wäre 120% statt 138,33% - oder denke ich da falsch?

Entfernt man die Steuern in der Dividenden-Buchung, zeigt der Vorsteuer-Graph korrekt 120% an. Der Fehler tritt also in der Beispieldatei nur mit der Dividende auf. Er lässt sich aber durch Steuerbuchungen, die nicht auf Kursgewinnen basieren und mit Steuerrückerstattungen ebenfalls verfälschen. Diese sollten ja eigentlich gar keinen Effekt auf den Vorsteuergraphen haben, haben ihn aber.

**Comment by buchen** [comment]
> Die kommer.xml habe ich nicht und finde sie auch nicht im Download-Bereich oder sonst irgendwo hier - wo bekomme ich sie?

Ich meinte damit die "eingebaute" Beispieldatei die man auf der "Willkommen" Seite öffnen kann.

> Der Gesamtportfolio-Chart unter Performance->Diagramm und die Berechnung nach Steuern sind korrekt. Der dort angezeigte Vorsteuer-Graph ist aber falsch. Richtig wäre 120% statt 138,33% - oder denke ich da falsch?

Okay, wie rechnet PP aktuell:

Erst die **Dividende**:

<img width="600" alt="bildschirmfoto 2019-01-25 um 10 21 17" src="https://user-images.githubusercontent.com/587976/51736690-081a6000-208b-11e9-8fd2-e38325cc734f.png">

Die Position ist mit 100 EUR bewertet, Du bekommst am 10.1. 30 Euro Dividende.
Macht +30%.
Allerdings verbleiben auf dem Konto ja wirklich nur 20 EUR, denn 10 EUR hast Du als Steuern bezahlt.
Das heißt PP bucht auch einen Performance-neutrale Entnahme von 10 Euro.
Also haben wir einen Bewertung von 100 Euro Aktien + 20 Euro Cash.

Dann kommt der **Verkauf**:

<img width="602" alt="bildschirmfoto 2019-01-25 um 10 21 23" src="https://user-images.githubusercontent.com/587976/51736843-6a736080-208b-11e9-8d6b-240f3438d276.png">

Der Verkauf bringt 200 Euro (vor Steuern). Macht 200 Euro + 20 auf dem Konto.
Macht ein Plus an diesem Tag von 220/120 = 1,8333333333
Damit kommt PP auf 1,3 * 1,833 = 2,3833 = 138,33%

Zusätzlich wird noch eine Performance-neutrale Entnahme von 50 Euro gebucht. Denn es sind ja nicht wirklich 200 Euro auf dem Konto angekommen sondern nur 150.

> Der dort angezeigte Vorsteuer-Graph ist aber falsch. Richtig wäre 120% statt 138,33% - oder denke ich da falsch?

Oder ich denke falsch. Wie sollte PP denn rechnen um auf die 120% (nach dem Verkauf) zu kommen?

**Comment by imperatom** [comment]
Oje, da hab ich die Fehelranalyse etwas vermasselt merke ich gerade - entschuldige bitte! Es lässt sich aber aufräumen und dabei hilft Deine Rechnung sehr:

Ich meinte eigentlich, dass ich **130**% für korrekt halten würde als Vorsteuerrendite (nämlich 30% aus der Dividende und 100% aus dem Verkauf).

Der Unterschied zu den von Dir berechenten 138,33% resultiert aus der performance-neutralen Entnahme von 10 Euro, weil sich die Bezugsgrößen dadurch ändern.

Du rechnest:
_Der Verkauf bringt 200 Euro (vor Steuern). Macht 200 Euro + 20 auf dem Konto.
Macht ein Plus an diesem Tag von 220/120 = 1,8333333333
Damit kommt PP auf 1,3 * 1,833 = 2,3833 (-100%) ->  138,33%_
Wenn man so rechnet, rechnet man sich die Dividendenbesteuerung aber durch die Hintertür wieder mit hinein, obwohl man ja eigentlich ohne Steuern rechnen wollte.

Ich halte es für sinnvoller, die steuerlichen Dividendeneffekte in einer Rechnung ohne Steuern ganz zu ignorieren und zu rechnen:
Der Verkauf bringt 200 Euro (vor Steuern). Macht 200 Euro + **30** aus der Dividende (auch wenn nur 20 davon auf dem Konto liegen, aber die Steuern werden hier ignoriert).
Macht ein Plus an diesem Tag von **230/130** = 1,769...
Damit kommt PP auf 1,3 * 1,769...= 2,3   (-100%) ->  130% - was ich als sinnvolles Ergebnis bezeichnen würde und der simplen Rechnung 100+30 entspricht.

Ist das verständlich?

**Comment by buchen** [comment]
> Wenn man so rechnet, rechnet man sich die Dividendenbesteuerung aber durch die Hintertür wieder mit hinein, obwohl man ja eigentlich ohne Steuern rechnen wollte.

Das ist ein absolut valider Punkt.

Mir ist noch nicht ganz klar was es bedeuten würde das Konto nicht auf dem "korrekten" Kontostand zu belassen. Nehmen wir an, man hat eine große Steuerrückerstattung und kauft davon weitere Papiere. Ohne Korrektur ist man plötzlich "geleveraged". Mir ist nicht ganz klar welche weiteren Effekte es noch geben könnte.

Wie auch immer: ich habe mal eine Version gebaut bei der Steuern nicht auf dem Konto korrigiert werden:

https://www.dropbox.com/s/eifr5wazgqlm4k0/PortfolioPerformance-distro-0.37.2-SNAPSHOT-win32.win32.x86_64.zip?dl=0
https://www.dropbox.com/s/7c4grtfdrkofua7/PortfolioPerformance-distro-0.37.2-SNAPSHOT-win32.win32.x86.zip?dl=0

**Comment by imperatom** [comment]
> Mir ist noch nicht ganz klar was es bedeuten würde das Konto nicht auf dem "korrekten" Kontostand zu belassen. Nehmen wir an, man hat eine große Steuerrückerstattung und kauft davon weitere Papiere. Ohne Korrektur ist man plötzlich "geleveraged". Mir ist nicht ganz klar welche weiteren Effekte es noch geben könnte.

Ja, wenn man nicht den korrekten Kontostand erfassen würde, bekäme man solche Fehlberechnungen, nicht nur bei Erstattungen, sondern auch bei Steuerzahlungen. Daher würde ich den Kontostand tatsächlich genau so veralten, wie Du es oben getan hast: Die Steuer wie eine performance-neutrale Entnahme betrachten - aber eben nur hinsichtlich der Liquidität, nicht hinsichtlich der Renditeberechnung. 
Gäbe es ein Problem mit dieser Unterscheidung? 

Unter Performance->Berechnung in PP benötigt man den Schalter "vor Steuern" eigentlich übrigens überhaupt nicht, denn hier werden ja gar keine prozentualen Renditen berechnet. Erst bei den Graphen im Diagramm sowie im Dashboard wird es interessant.

> Wie auch immer: ich habe mal eine Version gebaut bei der Steuern nicht auf dem Konto korrigiert werden:
> https://www.dropbox.com/s/eifr5wazgqlm4k0/PortfolioPerformance-distro-0.37.2-SNAPSHOT-win32.win32.x86_64.zip?dl=0

Prima, habe ich direkt mal ausprobiert mit meiner Testdatei von oben. Der Gesamtprofoliograph stimmt (70% nach Steuern), ebenso der Vorsteuer-Depot-Graph (130%). Für Aktie 1 wird mir jedoch 160% angezeigt, den Wert verstehe ich noch nicht.
